File: a2a-go-sdk/pubsub.go
package a2a

import (
	"context"
	"io"
	"log/slog"
	"math"
	"time"

	pb "github.com/dongchenglin0/a2a-framework/a2a-go-sdk/proto"
)

// PubSubSubscriber PubSub 订阅者，支持断线自动重连
type PubSubSubscriber struct {
	stub         pb.PubSubServiceClient
	agentID      string
	sessionToken string
	cancel       context.CancelFunc
}

// NewPubSubSubscriber 创建 PubSub 订阅者
func NewPubSubSubscriber(stub pb.PubSubServiceClient, agentID, sessionToken string) *PubSubSubscriber {
	return &PubSubSubscriber{
		stub:         stub,
		agentID:      agentID,
		sessionToken: sessionToken,
	}
}

// Subscribe 订阅指定 topics，消息到达时调用 handler，断线后自动重连
func (p *PubSubSubscriber) Subscribe(topics []string, handler PubSubHandler) {
	ctx, cancel := context.WithCancel(context.Background())
	p.cancel = cancel
	go p.subscribeLoop(ctx, topics, handler)
}

// subscribeLoop 订阅主循环，负责断线重连逻辑
func (p *PubSubSubscriber) subscribeLoop(ctx context.Context, topics []string, handler PubSubHandler) {
	attempt := 0
	for {
		// 检查是否已取消
		select {
		case <-ctx.Done():
			return
		default:
		}

		err := p.doSubscribe(ctx, topics, handler)

		// 如果是主动取消，直接退出
		if ctx.Err() != nil {
			slog.Info("PubSub 订阅已取消", "agentId", p.agentID)
			return
		}

		// 指数退避重连，最大等待 30 秒
		waitSec := math.Min(math.Pow(2, float64(attempt)), 30)
		wait := time.Duration(waitSec) * time.Second
		slog.Warn("PubSub 连接断开，准备重连",
			"wait", wait,
			"attempt", attempt,
			"error", err,
		)

		select {
		case <-ctx.Done():
			return
		case <-time.After(wait):
		}
		attempt++
	}
}

// doSubscribe 建立一次订阅流，持续接收消息直到断开
func (p *PubSubSubscriber) doSubscribe(ctx context.Context, topics []string, handler PubSubHandler) error {
	stream, err := p.stub.Subscribe(ctx, &pb.PubSubSubscribeRequest{
		AgentId:      p.agentID,
		Topics:       topics,
		SessionToken: p.sessionToken,
	})
	if err != nil {
		return err
	}

	slog.Info("PubSub 订阅已建立", "agentId", p.agentID, "topics", topics)

	for {
		msg, err := stream.Recv()
		if err == io.EOF {
			// 服务端正常关闭流
			return nil
		}
		if err != nil {
			return err
		}
		// 解析 payload 并调用处理器
		payload := StructToMap(msg.Payload)
		handler(msg.Topic, msg.PublisherAgentId, payload)
	}
}

// Unsubscribe 取消订阅，停止接收消息
func (p *PubSubSubscriber) Unsubscribe() {
	if p.cancel != nil {
		p.cancel()
	}
}

File: a2a-python-sdk/a2a/pubsub.py
"""
A2A SDK PubSub 订阅模块
提供带断线自动重连的异步 PubSub 订阅器
"""

import asyncio
import logging
from typing import Callable, Awaitable, Optional

import grpc

from a2a.proto import a2a_pb2, a2a_pb2_grpc

logger = logging.getLogger(__name__)

# PubSub 消息处理器类型：(topic, publisher_id, payload_dict) -> None
PubSubHandler = Callable[[str, str, dict], Awaitable[None]]


class PubSubSubscriber:
    """
    A2A PubSub 异步订阅器

    特性：
    - 订阅一个或多个 topic
    - 断线后指数退避自动重连
    - 支持优雅取消订阅

    用法::

        subscriber = PubSubSubscriber(channel, agent_id, session_token)
        await subscriber.subscribe(["topic.a", "topic.b"], my_handler)
        # ... 稍后 ...
        await subscriber.unsubscribe()
    """

    def __init__(
        self,
        channel: grpc.aio.Channel,
        agent_id: str,
        session_token: str,
    ):
        self._stub = a2a_pb2_grpc.PubSubServiceStub(channel)
        self._agent_id = agent_id
        self._session_token = session_token
        self._subscribe_task: Optional[asyncio.Task] = None
        self._running: bool = False

    async def subscribe(self, topics: list[str], handler: PubSubHandler) -> None:
        """
        开始订阅指定 topics，消息到达时调用 handler。
        断线后自动重连（指数退避，最大等待 30 秒）。

        :param topics: 要订阅的 topic 列表
        :param handler: 消息处理器 async (topic, publisher_id, payload) -> None
        """
        self._running = True
        self._subscribe_task = asyncio.create_task(
            self._subscribe_loop(topics, handler)
        )
        logger.info(f"PubSub 订阅已启动: agent_id={self._agent_id}, topics={topics}")

    async def _subscribe_loop(self, topics: list[str], handler: PubSubHandler) -> None:
        """
        订阅主循环，负责建立流式连接并处理消息。
        遇到连接错误时按指数退避重连。
        """
        attempt = 0

        while self._running:
            try:
                request = a2a_pb2.PubSubSubscribeRequest(
                    agent_id=self._agent_id,
                    topics=topics,
                    session_token=self._session_token,
                )

                logger.debug(f"PubSub 正在连接... (attempt={attempt})")

                # 建立服务端流式连接
                async for msg in self._stub.Subscribe(request):
                    if not self._running:
                        break

                    # 重置重试计数（说明连接正常）
                    attempt = 0

                    # 解析消息载荷
                    from a2a.utils import struct_to_dict
                    payload = struct_to_dict(msg.payload)

                    # 调用用户处理器
                    try:
                        await handler(msg.topic, msg.publisher_agent_id, payload)
                    except Exception as e:
                        logger.error(
                            f"PubSub handler 执行异常: topic={msg.topic}, error={e}",
                            exc_info=True,
                        )

            except asyncio.CancelledError:
                # 被主动取消，退出循环
                logger.debug("PubSub 订阅循环已取消")
                break

            except grpc.aio.AioRpcError as e:
                if not self._running:
                    break

                # 指数退避重连：1, 2, 4, 8, 16, 30, 30, ...
                wait = min(2 ** attempt, 30)
                logger.warning(
                    f"PubSub 连接断开，{wait}s 后重连... "
                    f"(attempt={attempt}, code={e.code()}, details={e.details()})"
                )
                try:
                    await asyncio.sleep(wait)
                except asyncio.CancelledError:
                    break
                attempt += 1

            except Exception as e:
                if not self._running:
                    break

                wait = min(2 ** attempt, 30)
                logger.error(
                    f"PubSub 发生未知错误，{wait}s 后重连... (attempt={attempt}): {e}",
                    exc_info=True,
                )
                try:
                    await asyncio.sleep(wait)
                except asyncio.CancelledError:
                    break
                attempt += 1

        logger.info(f"PubSub 订阅循环已退出: agent_id={self._agent_id}")

    async def unsubscribe(self) -> None:
        """
        取消订阅，停止重连循环。
        """
        self._running = False
        if self._subscribe_task and not self._subscribe_task.done():
            self._subscribe_task.cancel()
            try:
                await self._subscribe_task
            except asyncio.CancelledError:
                pass
        self._subscribe_task = None
        logger.info(f"PubSub 已取消订阅: agent_id={self._agent_id}")
